src/workers/parser/mapBindings.js

```diff
@@ -10,20 +10,51 @@ import { isTopLevel } from "./utils/helpers";
 import generate from "@babel/generator";
 import * as t from "@babel/types";
 
+function getIdentifierName(identifier, bindings) {
+  return bindings.includes(identifier.name)
+    ? identifier
+    : t.memberExpression(t.identifier("self"), identifier);
+}
+
+export function globalizeObjectDestructuring(node, bindings) {
+  for (const property of node.properties) {
+    if (t.isIdentifier(property.value)) {
+      property.value = getIdentifierName(property.value, bindings);
+    } else if (
+      t.isAssignmentPattern(property.value) &&
+      t.isIdentifier(property.value.left)
+    ) {
+      property.value.left = getIdentifierName(property.value.left, bindings);
+    }
+  }
+}
+
 // translates new bindings `var a = 3` into `self.a = 3`
 // and existing bindings `var a = 3` into `a = 3` for re-assignments
 function globalizeDeclaration(node, bindings) {
   return node.declarations.map(declaration => {
-    const identifier = bindings.includes(declaration.id.name)
-      ? declaration.id
-      : t.memberExpression(t.identifier("self"), declaration.id);
+    if (t.isPattern(declaration.id)) {
+      return t.expressionStatement(
+        t.assignmentExpression("=", declaration.id, declaration.init)
```

Comment:
I think ideally we'd want to call `globalizeObjectDestructuring` and `globalizeArrayDestructuring` from here as a general
```
t.assignmentExpression("=", globalizeDestructuring(declaration.id), declaration.init)
```
otherwise we're temporarily putting the AST into a broken state, since
```
var {a: self.a } = {}
```
is not valid syntax. It would also make things easier to follow in my opinion.

Reply:
We could also potentially instead change the `getIdentifierName(declaration.id, bindings);` below to `getAssignmentTarget(declaration.id, bindings)` what would automatically return the right thing to assign to instead.